stdlib/LinearAlgebra/src/triangular.jl

```diff
@@ -2435,25 +2435,31 @@ function _sqrt_quasitriu_offdiag_block!(R, A)
     return R
 end
 
+# real square root of 2x2 diagonal block of quasi-triangular matrix from real Schur
+# decomposition. Eqs 6.8-6.9 and Algorithm 6.5 of
+# Higham, 2008, "Functions of Matrices: Theory and Computation", SIAM.
 Base.@propagate_inbounds function _sqrt_real_2x2!(R, A)
     a11, a21, a12, a22 = A[1, 1], A[2, 1], A[1, 2], A[2, 2]
+    d = a11 - a22
     θ = (a11 + a22) / 2
-    μ² = -(a11 - a22)^2 / 4 - a21 * a12
-    μ = sqrt(μ²)
-    if θ > 0
-        α = sqrt((sqrt(θ^2 + μ²) + θ) / 2)
-    else
-        α = μ / sqrt(2 * (sqrt(θ^2 + μ²) - θ))
-    end
+    μ = sqrt(-d^2 - 4 * a21 * a12) / 2
```

Comment:
I updated this method to follow exactly what Higham recommended in the reference. Also following @stevengj's suggestion in https://github.com/JuliaLang/julia/pull/39973#discussion_r599035462, now `hypot` is used for computing `α`. @stevengj, can you clarify your suggestion on how `hypot` could be used to compute `μ` more stably?

Reply:
It's not about numerical stability, it's about avoiding spurious overflow/underflow.  For example, consider the case where `a21 = -(a12 = 1e300)`.

It's not exactly a `hypot` call, but it can maybe use a similar algorithm, e.g.:
```s
μₛ = max(abs(d), abs(a21), abs(a12))
μ = iszero(μₛ) ? float(μₛ) : μₛ * sqrt(-(d/μₛ)^2 - 4 * (a21/μₛ) * (a12/μₛ)) / 2
```

Reply:
(Unfortunately, all those additional divisions might have a non-negligible performance impact, since it sounds like this routine is performance-critical.)

Reply:
Maybe just `μₛ = max(abs(a21), abs(a12))` will suffice, since it looks like you must always have `|d| < max(|a21|, |a12|)` in this function (in order to have a real square root).

Reply:
Okay, thanks, I'll try that.

Reply:
I tested, and this _substantially_ improves the stability of the matrix square root.